# Welcome to our Cogs 108 Final Project! Our Group members are:

### Alvin Ngo (A12786420)
### Chen Huang (forgot his PID)
### James McDougall (A14947178) but not really 

In [1]:
import pandas as pd # we use pandas to 
import matplotlib.pyplot as plt

In [2]:
# you'll both have the download the data yourself. It takes about ten minutes
crime = pd.read_csv('/Users/alvin/Downloads/Crimes_-_2001_to_present.csv')

# First, here is some basic information and statistics about our data

In [3]:
crime.describe().loc['count']

ID                6823075.0
Beat              6823075.0
District          6823022.0
Ward              6208222.0
Community Area    6209545.0
X Coordinate      6750638.0
Y Coordinate      6750638.0
Year              6823075.0
Latitude          6750638.0
Longitude         6750638.0
Name: count, dtype: float64

In [4]:
print('Earliest Year:' + str(crime['Year'].min()) + '\tLatest Year: ' + str(crime['Year'].max()))

Earliest Year:2001	Latest Year: 2019


In [5]:
for column in crime:
    print(column)

ID
Case Number
Date
Block
IUCR
Primary Type
Description
Location Description
Arrest
Domestic
Beat
District
Ward
Community Area
FBI Code
X Coordinate
Y Coordinate
Year
Updated On
Latitude
Longitude
Location


In [4]:
# find number of Nans in each column
for column in crime:
    # note: some columns are text, so all will be Nan.
    print('Column Name: ' + str(column) + '\t\tNumber of NaNs: ' + str(crime[str(column)].isnull().sum().sum()))

Column Name: ID		Number of NaNs: 0
Column Name: Case Number		Number of NaNs: 4
Column Name: Date		Number of NaNs: 0
Column Name: Block		Number of NaNs: 0
Column Name: IUCR		Number of NaNs: 0
Column Name: Primary Type		Number of NaNs: 0
Column Name: Description		Number of NaNs: 0
Column Name: Location Description		Number of NaNs: 4620
Column Name: Arrest		Number of NaNs: 0
Column Name: Domestic		Number of NaNs: 0
Column Name: Beat		Number of NaNs: 0
Column Name: District		Number of NaNs: 53
Column Name: Ward		Number of NaNs: 614853
Column Name: Community Area		Number of NaNs: 613530
Column Name: FBI Code		Number of NaNs: 0
Column Name: X Coordinate		Number of NaNs: 72437
Column Name: Y Coordinate		Number of NaNs: 72437
Column Name: Year		Number of NaNs: 0
Column Name: Updated On		Number of NaNs: 0
Column Name: Latitude		Number of NaNs: 72437
Column Name: Longitude		Number of NaNs: 72437
Column Name: Location		Number of NaNs: 72437


In [5]:
null_crime = crime[crime['District'].isnull()]
num_row_null_district = len(null_crime)

# Cleaning the data

We have some rows with no District... we can't do much with them. So we're just going to drop them. There aren't that many, though (only 47 in our table with about 6 million rows), so it shouldn't have an impact. Also, remove all rows which are in district 31... this district isn't part of the regular list on the website and i can't find any info on them

In [6]:
original_num_rows = len(crime)
crime = crime[ crime['District'].isnull() == False ] # remove the rows with null district
crime

ID Case Number                    Date  \
0        7956475    HT188097  03/04/2011 10:52:00 AM   
1        7956476    HT185132  03/04/2011 12:00:00 PM   
2        7956477    HT187881  03/04/2011 09:00:00 AM   
3        7956479    HT187939  03/03/2011 08:00:00 AM   
4        7956480    HT187986  03/04/2011 10:35:00 AM   
5        7956485    HT188051  03/04/2011 12:00:00 AM   
6        7956486    HT188132  03/04/2011 11:46:00 AM   
7        7956488    HT188001  03/03/2011 10:40:00 PM   
8        7956490    HT188163  03/04/2011 10:00:00 AM   
9        7956494    HT188139  03/04/2011 12:05:00 PM   
10       7956495    HT187739  03/04/2011 07:20:00 AM   
11       7956496    HT188077  03/02/2011 08:35:00 PM   
12       7956497    HT187815  03/04/2011 08:40:00 AM   
13       7956501    HT188162  03/04/2011 12:20:00 PM   
14       7956502    HT187928  03/03/2011 07:00:00 PM   
15       7956503    HT187989  03/04/2011 10:00:00 AM   
16       7956505    HT188110  02/25/2011 07:15:00 PM   
17       7956506    HT184976  03/01/2011 11:30:00 AM   
18       7956507    HT188134  01/16/2011 09:00:00 AM   
19       7956512    HT188036  03/04/2011 05:00:00 AM   
20       7956513    HT188156  12/28/2010 09:30:00 AM   
21       7956515    HT188164  03/04/2011 11:30:00 AM   
22       7956517    HT188184  03/02/2011 02:00:00 AM   
23       7956518    HT185272  03/04/2011 12:00:00 PM   
24       7956525    HT188127  03/03/2011 03:00:00 PM   
25       7956526    HT188124  03/04/2011 12:15:00 PM   
26       7956527    HT187839  03/04/2011 09:26:00 AM   
27       7956528    HT188034  03/04/2011 06:00:00 AM   
28       7956532    HT188090  03/04/2011 11:11:00 AM   
29       7956535    HT169639  02/19/2011 11:11:00 AM   
...          ...         ...                     ...   
6823045  7956430    HT188061  03/02/2011 04:00:00 PM   
6823046  7956432    HS643742  12/02/2010 02:45:00 PM   
6823047  7956433    HT187855  12/22/2010 12:00:00 PM   
6823048  7956434    HT169823  02/15/2011 05:00:00 PM   
6823049  7956436    HT188147  03/03/2011 02:00:00 PM   
6823050  7956437    HT188014  03/04/2011 12:01:00 AM   
6823051  7956438    HT186493  02/28/2011 10:00:00 AM   
6823052  7956439    HT187895  03/04/2011 09:40:00 AM   
6823053  7956442    HT187968  03/04/2011 10:00:00 AM   
6823054  7956443    HT188095  03/04/2011 11:00:00 AM   
6823055  7956444    HT187953  03/03/2011 02:30:00 PM   
6823056  7956445    HT188112  03/03/2011 03:00:00 PM   
6823057  7956448    HT186436  03/03/2011 03:00:00 AM   
6823058  7956449    HT188142  03/04/2011 12:25:00 PM   
6823059  7956450    HT187952  03/03/2011 06:30:00 PM   
6823060  7956451    HT188116  03/04/2011 11:30:00 AM   
6823061  7956454    HT188169  03/04/2011 12:05:00 AM   
6823062  7956455    HT187831  03/03/2011 10:00:00 PM   
6823063  7956456    HT188069  03/04/2011 12:00:00 AM   
6823064  7956457    HT188009  03/04/2011 11:05:00 AM   
6823065  7956461    HT187040  03/04/2011 11:30:00 AM   
6823066  7956462    HT188153  03/04/2011 11:00:00 AM   
6823067  7956463    HT188115  03/04/2011 11:25:00 AM   
6823068  7956465    HT188074  03/04/2011 10:30:00 AM   
6823069  7956469    HT186678  05/10/2010 04:45:00 AM   
6823070  7956470    HT187756  03/02/2011 08:00:00 PM   
6823071  7956471    HT187826  03/04/2011 04:20:00 AM   
6823072  7956472    HT183976  03/04/2011 11:30:00 AM   
6823073  7956473    HT188121  03/04/2011 11:30:00 AM   
6823074  7956474    HT188024  03/04/2011 09:30:00 AM   

                                       Block  IUCR  \
0                        025XX S KILDARE AVE  0554   
1                            024XX W 45TH PL  1150   
2                       023XX N CLYBOURN AVE  1310   
3                           003XX W 100TH ST  0890   
4                      065XX W BRYN MAWR AVE  0460   
5                      042XX N HERMITAGE AVE  1150   
6                        043XX S PRAIRIE AVE  1811   
7                         041XX N WHIPPLE ST  0910   
8                     043XX W IRVING PARK RD  2851   
9

In [7]:
assert len(crime) == original_num_rows - num_row_null_district

Now, we want to see which primary types of crimes have happened the most and grab the top two

In [8]:
crime = crime[ crime['District'] <= 25 ]
crime['Primary Type'].value_counts()

THEFT                                1436417
BATTERY                              1245721
CRIMINAL DAMAGE                       778680
NARCOTICS                             715160
ASSAULT                               424099
OTHER OFFENSE                         423782
BURGLARY                              391088
MOTOR VEHICLE THEFT                   317092
DECEPTIVE PRACTICE                    268511
ROBBERY                               258175
CRIMINAL TRESPASS                     195337
WEAPONS VIOLATION                      72266
PROSTITUTION                           68529
PUBLIC PEACE VIOLATION                 48181
OFFENSE INVOLVING CHILDREN             46082
CRIM SEXUAL ASSAULT                    27700
SEX OFFENSE                            25482
INTERFERENCE WITH PUBLIC OFFICER       15522
GAMBLING                               14433
LIQUOR LAW VIOLATION                   14114
ARSON                                  11259
HOMICIDE                                9553
KIDNAPPING

We can see that the top two types are Theft (1436417
) and Battery (1245721
).
Therefore, we know want to see how many of each type actually concluded with an arrest.

In [9]:
print(crime[(crime['Primary Type']=='THEFT') & (crime['Arrest']==True)].shape)
print(crime[(crime['Primary Type']=='BATTERY') & (crime['Arrest']==True)].shape)
#print(crime[(crime['Primary Type']=='CRIMINAL DAMAGE') & (crime['Arrest']==True)].shape)

(171363, 22)
(283616, 22)


Now, we want to see the data in regards to our other possible variable (District). So we get the number of crimes for district first.

In [10]:
crime['District'].value_counts()

8.0     463806
11.0    437002
7.0     402525
25.0    393526
6.0     393392
4.0     386659
3.0     346487
9.0     338674
12.0    334279
2.0     323796
19.0    304132
5.0     302415
18.0    301186
15.0    297379
10.0    292151
14.0    268890
1.0     266914
16.0    225918
22.0    223523
24.0    203394
17.0    197798
20.0    118987
21.0         4
Name: District, dtype: int64

We can see that the top 2 districts with crimes are Districts 8.0 and 11.0. Therefore, we want to see out of all these reported crimes in these 2 districts, how many actually resulted in an arrest? 

In [12]:
print(crime[(crime['District']==8.0) & (crime['Arrest']==True)].shape)
print(crime[(crime['District']==11.0) & (crime['Arrest']==True)].shape)

(110888, 22)
(189984, 22)


In [11]:
assert crime['District'].max() != 31.0

# Run this cell to download the cleaned data to a csv in Unix...

In [39]:
crime.to_csv('~/Downloads/cleaned_chicago_crime_data.csv')

# Divide the dataframe into districts

In [36]:
district_list = crime['District'].unique().tolist()
district_list.sort()
district_list

[1.0,
 2.0,
 3.0,
 4.0,
 5.0,
 6.0,
 7.0,
 8.0,
 9.0,
 10.0,
 11.0,
 12.0,
 14.0,
 15.0,
 16.0,
 17.0,
 18.0,
 19.0,
 20.0,
 21.0,
 22.0,
 24.0,
 25.0]

In [37]:
districts = {}
for district in district_list:
    districts[str('District ' + str(district))] = crime[ crime['District'] == district ]